File: minimal_python/DataTypes/Matrix44.py
from math import atan2, cos, fabs, sin, sqrt
import math

from DataTypes.MathDefinitions import DEG2RAD, RAD2DEG
import robotcontrolapp_pb2


class Matrix44:
    # Initializes a 4x4 matrix data as a unit matrix
    def __init__(self):
        self._data = [
            1.0,
            0.0,
            0.0,
            0.0,
            0.0,
            1.0,
            0.0,
            0.0,
            0.0,
            0.0,
            1.0,
            0.0,
            0.0,
            0.0,
            0.0,
            1.0,
        ]

    # Constructor, copies values from GRPC matrix
    def FromGrpc(grpcMatrix: robotcontrolapp_pb2.Matrix44):
        if len(grpcMatrix.data) != 16:
            raise Exception(
                "could not initialize Matrix44, GRPC matrix had invalid element count "
                + len(grpcMatrix.data)
            )
        result = Matrix44()
        for i in range(16):
            result._data[i] = grpcMatrix.data[i]
        return result

    # Creates a GRPC matrix and copies the values
    def ToGrpc(self) -> robotcontrolapp_pb2.Matrix44:
        result = robotcontrolapp_pb2.Matrix44()
        for value in self._data:
            result.data.append(value)
        return result

    # Gets a value by row and column index
    def Get(self, row: int, column: int) -> float:
        return self._data[4 * row + column]

    # Sets a value by row and column index
    def Set(self, row: int, column: int, value: float):
        self._data[4 * row + column] = value

    # Gets the X position (in mm)
    def GetX(self) -> float:
        return self._data[3]

    # Gets the Y position (in mm)
    def GetY(self) -> float:
        return self._data[7]

    # Gets the Z position (in mm)
    def GetZ(self) -> float:
        return self._data[11]

    # Sets the X position (in mm)
    def SetX(self, x: float):
        self._data[3] = x

    # Sets the Y position (in mm)
    def SetY(self, y: float):
        self._data[7] = y

    # Sets the Z position (in mm)
    def SetZ(self, z: float):
        self._data[11] = z

    # Adds the given values to the position values (in mm)
    def Translate(self, x: float, y: float, z: float):
        self.SetX(self.GetX() + x)
        self.SetY(self.GetY() + y)
        self.SetZ(self.GetZ() + z)

    # Gets the A orientation (in degrees)
    def GetA(self) -> float:
        abc = self.GetOrientation()
        return abc[0]

    # Gets the B orientation (in degrees)
    def GetB(self) -> float:
        abc = self.GetOrientation()
        return abc[1]

    # Gets the C orientation (in degrees)
    def GetC(self) -> float:
        abc = self.GetOrientation()
        return abc[2]

    # Sets the A orientation (in degrees)
    def SetA(self, a: float):
        self.SetOrientation(a, self.GetB(), self.GetC())

    # Sets the A orientation (in degrees)
    def SetB(self, b: float):
        self.SetOrientation(self.GetA(), b, self.GetC())

    # Sets the A orientation (in degrees)
    def SetC(self, c: float):
        self.SetOrientation(self.GetA(), self.GetB(), c)

    # Gets the orientation (as 3-element tuple in degrees)
    def GetOrientation(self) -> tuple[float, float, float]:
        eps = 0.001
        b = atan2(
            -self._data[8],
            sqrt((self._data[0] * self._data[0] + self._data[4] * self._data[4])),
        )

        if fabs(b - math.pi / 2.0) < eps:  # singularity b = Pi/2
            a = 0.0
            c = atan2(self._data[1], self._data[5])
        elif fabs(b + math.pi / 2.0) < eps:  # singularity b = - PI/2
            a = 0.0
            c = -atan2(self._data[1], self._data[5])
        else:  # normal case
            cb = cos(b)
            a = atan2(self._data[4] / cb, self._data[0] / cb)
            c = atan2(self._data[9] / cb, self._data[10] / cb)

        a *= RAD2DEG
        b *= RAD2DEG
        c *= RAD2DEG

        return (a, b, c)

    # Sets the orientation (in degrees)
    def SetOrientation(self, a: float, b: float, c: float):
        alpha = a * DEG2RAD
        beta = b * DEG2RAD
        gamma = c * DEG2RAD

        sa = sin(alpha)
        ca = cos(alpha)
        sb = sin(beta)
        cb = cos(beta)
        sg = sin(gamma)
        cg = cos(gamma)

        self._data[0] = ca * cb
        self._data[1] = ca * sb * sg - sa * cg
        self._data[2] = ca * sb * cg + sa * sg
        self._data[4] = sa * cb
        self._data[5] = sa * sb * sg + ca * cg
        self._data[6] = sa * sb * cg - ca * sg
        self._data[8] = -sb
        self._data[9] = cb * sg
        self._data[10] = cb * cg

File: control_cpp/src/DataTypes/Matrix44.cpp
#include "Matrix44.h"

#include <exception>

#include "MathDefinitions.h"

namespace App
{
namespace DataTypes
{

/**
 * @brief Constructor, initializes a 4x4 unit matrix
 */
Matrix44::Matrix44()
{
    // initialize to unit matrix
    m_data.fill(0);
    m_data[0] = 1;
    m_data[5] = 1;
    m_data[10] = 1;
    m_data[15] = 1;
}

/**
 * @brief Constructor, copies values from GRPC matrix
 * @param other
 */
Matrix44::Matrix44(const robotcontrolapp::Matrix44& other)
{
    if (other.data().size() != (int)m_data.size()) throw std::runtime_error("could not initialize Matrix44, GRPC matrix had invalid element count " + std::to_string(m_data.size()));
    std::copy_n(other.data().data(), m_data.size(), m_data.data());
}

/**
 * @brief Element access operator
 * @param row row number (0 - 3), throws exception of out of range
 * @param column column number (0 - 3), throws exception of out of range
 * @return matrix element value
 */
double& Matrix44::operator()(int row, int column)
{
    if (row < 0 || row >= 4 || column < 0 || column >= 4) throw std::out_of_range("matrix index out of range");
    return m_data[4 * row + column];
}

/**
 * @brief Element access operator
 * @param row row number (0 - 3), throws exception of out of range
 * @param column column number (0 - 3), throws exception of out of range
 * @return matrix element value
 */
double Matrix44::operator()(int row, int column) const
{
    if (row < 0 || row >= 4 || column < 0 || column >= 4) throw std::out_of_range("matrix index out of range");
    return m_data[4 * row + column];
}

/**
 * @brief Adds the given values to the position values
 * @param x X position in mm
 * @param y Y position in mm
 * @param z Z position in mm
 */
void Matrix44::Translate(double x, double y, double z)
{
    SetX(GetX() + x);
    SetY(GetY() + y);
    SetZ(GetZ() + z);
}

/**
 * @brief Gets the A orientation
 * @return A orientation value
 */
double Matrix44::GetA() const
{
    double a = 0, b = 0, c = 0;
    GetOrientation(a, b, c);
    return a;
}

/**
 * @brief Gets the B orientation
 * @return B orientation value
 */
double Matrix44::GetB() const
{
    double a = 0, b = 0, c = 0;
    GetOrientation(a, b, c);
    return b;
}

/**
 * @brief Gets the C orientation
 * @return C orientation value
 */
double Matrix44::GetC() const
{
    double a = 0, b = 0, c = 0;
    GetOrientation(a, b, c);
    return c;
}

/**
 * @brief Sets the A orientation
 * @param a A angle in degrees
 */
void Matrix44::SetA(double a)
{
    SetOrientation(a, GetB(), GetC());
}

/**
 * @brief Sets the B orientation
 * @param b B angle in degrees
 */
void Matrix44::SetB(double b)
{
    SetOrientation(GetA(), b, GetC());
}

/**
 * @brief Sets the C orientation
 * @param c C angle in degrees
 */
void Matrix44::SetC(double c)
{
    SetOrientation(GetA(), GetB(), c);
}

/**
 * @brief Gets the orientation
 * @param a A angle in degrees
 * @param b B angle in degrees
 * @param c C angle in degrees
 */
void Matrix44::GetOrientation(double& a, double& b, double& c) const
{
    constexpr double eps = 0.001;
    b = atan2(-m_data[8], sqrt((m_data[0] * m_data[0] + m_data[4] * m_data[4])));

    if (fabs(b - M_PI / 2.0) < eps) // singularity b = Pi/2
    {
        a = 0.0;
        c = atan2(m_data[1], m_data[5]);
    }
    else if (fabs(b + M_PI / 2.0) < eps) // singularity b = - PI/2
    {
        a = 0.0;
        c = -atan2(m_data[1], m_data[5]);
    }
    else // normal case
    {
        double cb = cos(b);
        a = atan2(m_data[4] / cb, m_data[0] / cb);
        c = atan2(m_data[9] / cb, m_data[10] / cb);
    }

    a *= rad2deg;
    b *= rad2deg;
    c *= rad2deg;
}

/**
 * @brief Sets the orientation
 * @param a A angle in degrees
 * @param b B angle in degrees
 * @param c C angle in degrees
 */
void Matrix44::SetOrientation(double a, double b, double c)
{
    double alpha = a * deg2rad;
    double beta = b * deg2rad;
    double gamma = c * deg2rad;

    double sa = sin(alpha);
    double ca = cos(alpha);
    double sb = sin(beta);
    double cb = cos(beta);
    double sg = sin(gamma);
    double cg = cos(gamma);

    m_data[0] = ca * cb;
    m_data[1] = ca * sb * sg - sa * cg;
    m_data[2] = ca * sb * cg + sa * sg;
    m_data[4] = sa * cb;
    m_data[5] = sa * sb * sg + ca * cg;
    m_data[6] = sa * sb * cg - ca * sg;
    m_data[8] = -sb;
    m_data[9] = cb * sg;
    m_data[10] = cb * cg;
}

/**
 * @brief Creates a GRPC matrix and copies the values
 * @return GRPC matrix
 */
robotcontrolapp::Matrix44 Matrix44::ToGrpc() const
{
    robotcontrolapp::Matrix44 result;
    for (double val : m_data) result.add_data(val);
    return result;
}

} // namespace DataTypes
} // namespace App
